### PR TITLE
 Refactor most of the Minting contract log parsing and handling code 

### PR DESCRIPTION
**Pull request by stevenroose**

### CODE REVIEW

**Comment by stevenroose** [comment]
If you want to read some code that is actually readable, you can always review the second commit, which contains the bulk of this change, and then verify that the last commit only does `make fmt` ;)

File: contracts/src/Minting.sol [hunk 1]
```diff
@@ -6,18 +6,34 @@ contract Minting {
     uint constant GAS_SIMPLE_TRANSFER = 21000;
     uint constant GAS_AMOUNT_UPDATE = 2003;
     uint constant GAS_REVERT_TRUE = 3;
-    uint constant GAS_MINT_AMOUNT_EVENT = 1133;
+
```
Comment (arminsabouri) [inline contracts/src/Minting.sol:9]: Im probably not the best person to review this part
tagging @scottmillner to take a look

File: contracts/src/Minting.sol [hunk 2]
```diff
@@ -6,18 +6,34 @@ contract Minting {
...
+    // > indexed parameter, so that cost is included most of the time. Then we
+    // > pay 8 times the number of 32 byte words written to the chain. Because
+    // > this region is stored in memory before being emitted, the memory
+    // > expansion cost must be accounted for also.
+    uint constant BASE_GAS_MINT_EVENT = 1157;
 
     /// Some new coins have been minted.
     event Mint(
         address indexed account,
         uint256 amount,
```
Comment (arminsabouri) [inline contracts/src/Minting.sol]: Do we use amount anywhere? Seems like mint amount should suffice

File: crates/consensus/authority/src/block_builder.rs [hunk 3]
```diff
@@ -181,22 +185,41 @@ where
...
+            for receipt in receipts {
+                if idx == 0 && receipt.is_none() {
+                    break; // Prunning block, skip
+                }
+                if let Some(receipt) = receipt {
+                    if !receipt.success {
+                        continue;
+                    }
+                    for log in &receipt.logs {
+                        let pegin_match = try_parse_mint_event(log).expect("passed EVM check");
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs:194]: Comment justifiying the expect
```suggestion
                        // Mint event should have already been validated in during evm execution (in processor.rs) 
                        let pegin_match = try_parse_mint_event(log).expect("passed EVM check");
```

File: crates/consensus/authority/src/block_builder.rs [hunk 4]
```diff
@@ -181,22 +185,41 @@ where
...
+                            info!(target: "consensus::authority", "Parsing and sending minting event to btc_server");
+                            for pegin in &pegin_data.meta {
+                                //TODO(stevenroose) should this happen here?
+                                if let Err(e) = call_notify_pegin(&mut self.btc_server, pegin).await
+                                {
+                                    error!(target: "consensus::authority", ?e, "failed to notify btc_server of pegin");
+                                    return;
+                                }
+                                info!(target: "consensus::authority", "notifying btc server about pegin utxo");
+                            }
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs:205]: Should we continue here?
There is no reason to try to parse a burn event if already have a pegin.
Or maybe a if elseif

File: crates/consensus/authority/src/block_fetcher.rs [hunk 5]
```diff
@@ -206,26 +208,47 @@ where
...
+                                                //TODO(stevenroose) should this happen here?
+                                                if let Err(e) =
+                                                    call_notify_pegin(btc_server, &pegin).await
+                                                {
+                                                    error!(target: "consensus::authority", ?e, "failed to notify btc_server of pegin");
+                                                    return;
+                                                }
+                                                info!(target: "consensus::authority", "notifying btc server about pegin utxo");
+                                            }
+                                        }
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_fetcher.rs:237]: Same comment here?

File: crates/consensus/authority/src/block_fetcher.rs [hunk 6]
```diff
@@ -206,26 +208,47 @@ where
...
+                                    if !receipt.success {
+                                        continue;
+                                    }
+                                    for log in &receipt.logs {
+                                        let pegin_match =
+                                            try_parse_mint_event(log).expect("passed EVM check");
+                                        if let Some(pegin_data) = pegin_match {
+                                            info!(target: "consensus::authority", "Parsing and sending minting event to btc_server");
+                                            for pegin in &pegin_data.meta {
+                                                //TODO(stevenroose) should this happen here?
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_fetcher.rs:228]: Not a req for change.
I agree with this. Duplicate code. Ideally we have some Pegin/Pegout Processor util or struct

File: crates/revm/src/processor.rs [hunk 7]
```diff
@@ -221,86 +220,60 @@ where
...
         result: &ExecutionResult,
         botanix_consensus_pkg: Option<BotanixConsensusPackage>,
-        sender_address: Address, // used for reverted pegouts
-        sender_value: U256,      // used for reverted pegouts
-    ) -> Result<(), BlockExecutionError> {
+    ) -> Result<(), MintContractError> {
         let consensus_pkg = botanix_consensus_pkg.as_ref();
+        let btc_network = consensus_pkg
+            .map(|package| package.btc_network)
+            .unwrap_or_else(|| bitcoin::Network::Regtest);
```
Comment (arminsabouri) [inline crates/revm/src/processor.rs]: Maybe we should just panic here if no network is defined. If you are processing the botanix logs then this should be defined

File: crates/revm/src/processor.rs [hunk 8]
```diff
@@ -221,86 +220,60 @@ where
...
-                                pegin_data.account,
-                                pegin_data.amount,
-                                MintContractErrorType::Pegin.to_string(),
+            let pegin_data = match try_parse_mint_event(log)? {
+                None => continue,
+                Some(p) => p,
+            };
+
+            let bitcoin_checkpoint = consensus_pkg.expect("is some").bitcoin_checkpoint;
+            if pegin_data.bitcoin_block_height >= bitcoin_checkpoint.1 {
```
Comment (arminsabouri) [inline crates/revm/src/processor.rs]: Is this correct? The error returned in the height is below the checkpoint. But here we are checking if the pegin block height is great than or eq to the checkpoint.

File: crates/revm/src/processor.rs [hunk 9]
```diff
@@ -221,86 +220,60 @@ where
...
+                        bitcoin_checkpoint.1, pegin_data.bitcoin_block_height,
+                    ),
+                    revert_address: pegin_data.account,
+                    revert_amount: pegin_data.amount,
+                });
+            }
+            let aggregate_public_key = consensus_pkg.expect("is some").aggregate_public_key;
+            match pegin_data.validate(&bitcoin_checkpoint, &aggregate_public_key) {
+                Ok(aggregate_value) => {
+                    if aggregate_value == pegin_data.amount {
```
Comment (arminsabouri) [inline crates/revm/src/processor.rs]: Should this be mint_amount? Or are those the same thing now?

File: crates/consensus/authority/src/block_builder.rs (hunk 4 above)
Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs:205]: i guess you could have a contract that batches both pegins and pegouts in a single transaction so there could be pegin and pegout receipts in one tx...that is technically possible

File: crates/revm/src/processor.rs (hunk 9 above)
Comment (scottmillner) [reply crates/revm/src/processor.rs]: it's the same thing now

`Minting.sol` has been refactored to only emit a single event that contains both the original amount and the final mint amount

`try_parse_mint()` will set `PeginData.amount` to the final mint amount

and amount will be less than aggregate value bc tx cost is deducted from original amount

File: contracts/src/Minting.sol (hunk 2 above)
Comment (scottmillner) [reply contracts/src/Minting.sol]: we don't but it's cheap to include in the event and my preference is to keep it
makes it very transparent if anyone really wanted to dig through logs and make sure everything is above board

**Comment by scottmillner** [comment]
acking it all looks good
just needs an error message to be updated...see armin's comment

File: crates/revm/src/processor.rs (hunk 8 above)
Comment (scottmillner) [reply crates/revm/src/processor.rs]: the check is correct but the error message isn't

- `bitcoin_checkpoint` is the header at the required depth (`let height = height.saturating_sub(pegin_conf_depth as usize - 1);`)
- so the pegin `bitcoin_block_height` should be equal or less than the checkpoint

ive updated the error message to reflect that and added a comment to explain

File: crates/revm/src/processor.rs (hunk 7 above)
Comment (scottmillner) [reply crates/revm/src/processor.rs]: done!

File: contracts/src/Minting.sol (hunk 2 above)
Comment (scottmillner) [reply contracts/src/Minting.sol]: actually forgot you can only have 4 topics so will just use mint
that's why the int tests were failing: the topic hash was incorrect since the last topic wasn't being included

**Review by arminsabouri — CHANGES_REQUESTED** [review]
tNack
getting this on a pegin
```
Botanix Minting contract event validation failed: InvalidPeginData { error: "pegin height 350 greater than checkpoint of 353", revert_address: 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266, revert_amount: 9999932007753125000 }
```

Not sure why the int. test is passing

**Comment by arminsabouri** [comment]
> tNack getting this on a pegin
> 
> ```
> Botanix Minting contract event validation failed: InvalidPeginData { error: "pegin height 350 greater than checkpoint of 353", revert_address: 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266, revert_amount: 9999932007753125000 }
> ```
> 
> Not sure why the int. test is passing

I believe the fix needed for this is here: https://github.com/botanix-labs/Macbeth/pull/314

When I manually add that check my pegin passes

**Review by rwlockbg — DISMISSED** [review]

File: crates/revm/src/processor.rs [hunk 10]
```diff
@@ -221,93 +220,69 @@ where
...
+                        pegin_data.bitcoin_block_height, bitcoin_checkpoint.1,
+                    ),
+                    revert_address: pegin_data.account,
+                    revert_amount: pegin_data.amount,
+                });
+            }
+            let aggregate_public_key = consensus_pkg.expect("is some").aggregate_public_key;
+            match pegin_data.validate(&bitcoin_checkpoint, &aggregate_public_key) {
+                Ok(aggregate_value) => {
+                    if aggregate_value == pegin_data.amount {
```
Comment (arminsabouri) [inline crates/revm/src/processor.rs]: `if aggregate_value == pegin_data.amount`
Cab you explain the logic here? Is it due to the refund? 
Is it possible that a refund reciever isnt defined and the refund never occurs?

Comment (arminsabouri) [reply crates/revm/src/processor.rs]: Additionally what is its greater? Should this be >= ?

File: crates/revm/src/processor.rs [hunk 11]
```diff
@@ -369,77 +344,56 @@ where
...
-                            };
-                            let new_result = ExecutionResult::Revert {
-                                gas_used: result.gas_used(),
-                                output: output_match,
-                            };
-
+                    match Self::botanix_mint_contract_checks(&result, botanix_consensus_pkg) {
+                        Ok(()) => Ok(ResultAndState { result, state }),
+                        Err(e) => {
+                            error!("Botanix Minting contract event validation failed: {:?}", e);
```
Comment (arminsabouri) [inline crates/revm/src/processor.rs:354]:
```suggestion
                            error!("Botanix Minting contract burn event validation failed: {:?}", e);
```

File: crates/revm/src/processor.rs [hunk 12]
```diff
@@ -369,77 +344,56 @@ where
...
-                                } else {
-                                    let mut updated_state = state.clone();
+                                }
+                                MintContractError::InvalidPegoutData(_) => {
                                     Self::increment_balance_by_address(
-                                        address,
-                                        amount,
-                                        &mut updated_state,
+                                        sender,
+                                        EthersU256::from_little_endian(pegout_amount.as_le_slice()),
```
Comment (arminsabouri) [inline crates/revm/src/processor.rs:373]: Should gas be account for here?

Comment (arminsabouri) [reply crates/revm/src/processor.rs:373]: Otherwise if someone can keep calling burn without having some account balance decremented won't that create supply that doesnt exist?

File: crates/revm/src/processor.rs (hunk 10 above)
Comment (scottmillner) [reply crates/revm/src/processor.rs]: yes, it is do to the refund so the aggregate amount should always be lower than the `amount` passed to the contract call

the `refundAddress` has to be passed to the contract call since it's part of the method signature, if it's not passed the call will fail

yes, it should be `>=`...nice catch!
done!

File: crates/revm/src/processor.rs (hunk 12 above)
Comment (scottmillner) [reply crates/revm/src/processor.rs:373]: gas is paid by the sender of the call to `mint()` which is sidecar at the moment but could be anyone

the EVM enforces that you have enough funds to cover gas and that has already been decremented from the sender account since this is a valid tx it terms of the EVM. It's just our additional logic that reverts it.  So gas has been handled correctly by the EVM before it gets to our custom validation logic.  All our logic needs to do is decrement the amount that was transferred to the destination address which we parse from `emit Mint(destination, amount, bitcoinBlockHeight, metadata);` in the call method.

TL;DR
it's impossible to call `burn()` without having enough ether to cover gas due to normal EVM rules

File: crates/revm/src/processor.rs (hunk 10 above)
Comment (arminsabouri) [reply crates/revm/src/processor.rs]: Nice!

**Comment by arminsabouri** [comment]
@scottmillner Do you mind running thru some manual pegin/pegouts on this PR? As well as the functional test suite?
